README.md
# flash-messenger
A jQuery plugin provides a simple flash messages and interface.

## Demo
[Default](https://emn178.github.io/flash-messenger/samples/default/)  
[Custom](https://emn178.github.io/flash-messenger/samples/bootstrap/)  
[Switch Methods](https://emn178.github.io/flash-messenger/samples/methods/)

## Download
[Compress JS](https://raw.github.com/emn178/flash-messenger/master/build/flash-messenger.min.js)  
[Compress CSS](https://raw.github.com/emn178/flash-messenger/master/build/flash-messenger.min.css)  
[Uncompress JS](https://raw.github.com/emn178/flash-messenger/master/src/flash-messenger.js)  
[Uncompress CSS](https://raw.github.com/emn178/flash-messenger/master/src/flash-messenger.css)

## Installation
You can also install flash-messenger by using Bower.
```
bower install flash-messenger
```

## Requirements

[jQuery 1.8+](http://jquery.com/)

## Usage
```JavaScript
flash('Messange');
flash('Messange', options);
```

### Options
#### *type: string (default: `success`)*

Sets the message type, it could be: 'error', 'danger', 'info', 'notice', 'success', 'warning', 'alert' in default theme. You can call methods with the same name like `flash.error()`, `flash.alert()`...

#### *time: number (default: `3000`)*

Sets how long the message stay in seconds.

#### *sticky: boolean (default: `false`)*

Sets if the message should be sticky. The default value will be true if type is 'error' or 'danger'.

#### *fadeOut: number (default: `1000`)*

Sets the duration to fade out in seconds.

#### *closable: boolean (default: `true`)*

Sets the message can be closed or not.

#### *scrollTo: boolean (default: `true`)*

Scroll to message position if true.

#### *html: boolean (default: `false`)*

Sets the message if it can be html.

### Default Setting
You can change default setting.
```JavaScript
flash.setting = {
  appendTo: 'body',
  container: '<div class="flash-messages"></div>',
  message: '<div class="flash-message {type}">{message}<span class="flash-message-close">&#10006</span></div>',
  closeHandler: '.flash-message-close',
  default: {
    type: 'success',
    time: 3000,
    sticky: false,
    fadeOut: 1000,
    closable: true,
    scrollTo: true,
    html: false
  },
  typesDefault: {
    error: {
      sticky: true
    },
    danger: {
      sticky: true
    }
  }
};
```

## Example
```JavaScript
flash('Messange');
flash('Messange', {type: 'error'});
flash.error('Message');
flash.info('Message', {sticky: true});
```

## Custom
```JavaScript
flash.setting.method = function (message, options) {
  // overwrite implementation
};
```

## License
The project is released under the [MIT license](http://www.opensource.org/licenses/MIT).

## Contact
The project's website is located at https://github.com/emn178/flash-messenger  
Author: Chen, Yi-Cyuan (emn178@gmail.com)


CHANGELOG.md
# Change Log

## v0.3.0 / 2019-05-06
### Added
- support turbolinks 5+.

## v0.2.1 / 2017-10-18
### Fixed
- not working above jquery 3+.

## v0.2.0 / 2017-10-02
### Added
- html option.

## v0.1.1 / 2016-05-02
### Added
- scrollTo option.

## v0.1.0 / 2016-04-13
### Added
- create project


src/flash-messenger.js
/*!
 * [flash-messenger]{@link https://github.com/emn178/flash-messenger}
 *
 * @version 0.3.0
 * @author Chen, Yi-Cyuan [emn178@gmail.com]
 * @copyright Chen, Yi-Cyuan 2016-2019
 * @license MIT
 */
(function ($) {
  'use strict';

  var OPTIONS = ['type', 'time', 'sticky', 'fadeOut', 'closable', 'scrollTo', 'html'];
  var TYPES = ['error', 'danger', 'info', 'notice', 'success', 'warning', 'alert'];

  var setting = {
    appendTo: 'body',
    container: '<div class="flash-messages"></div>',
    message: '<div class="flash-message {type}">{message}<span class="flash-message-close">&#10006</span></div>',
    closeHandler: '.flash-message-close',
    default: {
      type: 'success',
      time: 3000,
      sticky: false,
      fadeOut: 1000,
      closable: true,
      scrollTo: true,
      html: false
    },
    typesDefault: {
      error: {
        sticky: true
      },
      danger: {
        sticky: true
      }
    }
  };

  var container, encorder = $('<div/>'), screenHeight, screenWidth;
  function htmlEncode (message) {
    return encorder.text(message).html();
  }

  function formatOptions(options) {
    var typesDefault = setting.typesDefault[options.type];
    if (typesDefault) {
      $.extend(options, $.extend({}, typesDefault, options));
    }
    OPTIONS.forEach(function (key) {
      var value = options[key];
      options[key] = value === undefined ? setting.default[key] : value;
    });
    if (setting.convert) {
      setting.convert(options);
    }
  }

  function show(message, options) {
    options = options || {};
    formatOptions(options);
    if (!options.html) {
      message = htmlEncode(message);
    }
    var div = $(setting.message.replace('{message}', message).replace('{type}', options.type)).appendTo(container);
    var close = function () {
      div.remove();
    };
    if (!options.sticky) {
      div.delay(options.time).fadeOut(options.fadeOut, close);
    }
    if (options.closable) {
      div.find(setting.closeHandler).addBack(setting.closeHandler).bind('click', close);
    } else {
      div.addClass('nonclosable');
    }
    if (options.scrollTo) {
      var rect = div[0].getBoundingClientRect();
      if (rect.top < 0 || rect.top > screenHeight) {
        div[0].scrollIntoView();
      }
    }
  }

  function resize() {
    screenHeight = window.innerHeight || document.documentElement.clientHeight;
    screenWidth = window.innerWidth || document.documentElement.clientWidth;
  }

  var flash = function () {
    setting.method.apply(this, arguments);
  };
  flash.setting = setting;
  flash.formatOptions = formatOptions;
  setting.method = flash.defaultMethod = show;
  window.flash = flash;

  TYPES.forEach(function (type) {
    flash[type] = function (message, options) {
      options = options || {};
      options.type = type;
      flash(message, options);
    }
  });

  $(window).resize(resize);

  function load() {
    if (!container) {
      container = $(setting.container);
    }
    $(setting.appendTo).append(container);
    if (setting.window) {
      window.flash = setting.window.flash;
    }
  }

  $(document).ready(resize).ready(load).bind('page:load turbolinks:load', load);
})(jQuery);
